flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp

```diff
@@ -154,6 +156,56 @@ mlir::LLVM::DITypeAttr DebugTypeGenerator::convertBoxedSequenceType(
       dataLocation, /*rank=*/nullptr, allocated, associated);
 }
 
+mlir::LLVM::DITypeAttr DebugTypeGenerator::convertRecordType(
+    fir::RecordType Ty, mlir::LLVM::DIFileAttr fileAttr,
+    mlir::LLVM::DIScopeAttr scope, mlir::Location loc) {
+  mlir::MLIRContext *context = module.getContext();
+  auto result = fir::NameUniquer::deconstruct(Ty.getName());
+  if (result.first != fir::NameUniquer::NameKind::DERIVED_TYPE)
+    return genPlaceholderType(context);
+
+  std::optional<mlir::DataLayout> dl =
+      fir::support::getOrSetDataLayout(module, /*allowDefaultLayout=*/true);
```

Comment:
Building a datalayout is not cheap and leverage caching to speed things up (contains many DenseMap, requires doing parsing of module attributes): it should be done only once at the pass level.

https://github.com/llvm/llvm-project/blob/af1d2b9fb1c6065d3a4611f086b8c83a0c05795c/mlir/docs/DataLayout.md?plain=1#L84

Reply:
Thanks for the suggestion. I have now updated the code to initialize `DataLayout` once and use it wherever needed.